llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp

```diff
@@ -5234,6 +5234,20 @@ SDValue DAGCombiner::visitAVG(SDNode *N) {
   if (N0 == N1 && Level >= AfterLegalizeTypes)
     return N0;
 
+  // Fold avgflooru(x,c) -> avgceilu(x,c-1)
+  // iff c > 0 and avgflooru isn't legal/custom but avgceilu is.
```

Comment:
Is this a quirk of some specific target? Otherwise, why not do this for the signed cases too, and do the opposite fold from ceil -> floor?

Reply:
Its purely to get around x86 regressions (which only has avgceilu) - we can add other equivalences if there are other targets don't have all the avg variants.

Reply:
> Its purely to get around x86 regressions (which only has avgceilu)

Ack. Is it worth legalizing the signed versions to the unsigned ones by flipping the sign bit of both inputs and the result?

Reply:
Yes, that's another one that is probably worth it (not sure if we have an existing test case for that, although the avg tests are more extensive than you'd expect.....). These are pretty generic folds, but I can move this into X86 if people prefer?

Reply:
Unless there is certainty that no other target can benefit, imo its written generically enough to keep here.

I.e the general pattern: `if(NotLegal(A) && IsLegal(B) && MayTransform(A, B)) { return B; }` is sensible as a generic combine, even if the motivation was one particular target.